### DISCUSSION TITLE
How to know the controller class in a KernelEvent::RESPONSE subscriber?

### DISCUSSION THREAD

**Discussion by Pixelshaped** [symfony/symfony#50353] [Q&A] [answered]
I've recently written an event subscriber for a KernelEvent::RESPONSE subscriber.

I needed the response to be existing to delete some cookies in some case. I would have liked this deletion to be scoped to some controller classes though, and I found that, contrary to `ControllerEvent` or `ControllerArgumentsEvent`, there was no way to know the Controller class linked to this event.

Which seemed like a shame because it's right there, in the caller `HttpKernel::handleRaw()`

I guess I could inject a ControllerResolver in my event subscriber, the same it's done in `handleRaw`.

But it seems a bit convoluted. Is there a good reason that the controller class isn't passed as a parameter to all events fired south of `handleRaw:146`(this line: `if (false === $controller = $this->resolver->getController($request)) {`)

If not, I could eventually provide a PR.

Good day!

**Comment by stof** [comment]
the kernel.response event is triggered for all responses, not only the ones coming from a controller (listeners on kernel.request can provide a response for instance)

**Reply by Pixelshaped** [reply]
Thanks for your quick reply! In those cases couldn't the getter `getController()` that would be added to those events just return `null`?

**Comment by ahundiak** [comment]
Having the response depending on the controller seems a bit iffy but if you really want it then just add a controller listener and have it store the controller into the request as a custom attribute.

**Reply by Pixelshaped** [reply]
Iffy I don't know. Imagine you have a `ShoppingController` with a bunch of routes, and you want a response header set up for all routes declared in that controller. Wouldn't it make sense to be able to do that without having to rely on another listener and a custom attribute?

**Comment by mbabker** [comment]
Would the `_route` attribute from the request not work for your use case?

**Reply by Pixelshaped** [reply]
Performance of `$controller instanceof ShoppingController` is certainly superior to comparing a string route prefix or worse, to find out the Controller based on the route. But it could be a workaround you are right.

**Reply by mbabker** [reply, marked as answer]
As was pointed out earlier, there won't always be a controller if something sets a response early.  Nor is it a hard requirement that a controller be a method inside a class (though that's a pretty standard convention).

Worrying about the performance of `instanceof` versus `str_starts_with()` or `in_array($route, [], true)` seems like a premature optimization concern IMO.  And once you factor in having to look at `$event->getController()` to work out the type of callable you're working with (is it an array, a callable object, a string, a Closure?) any performance concerns go out the window.

There is one other workaround if you really need the resolved callable; listen on the `kernel.controller` event and store a reference to it in your subscriber and use that in your `kernel.response` listener.

Beyond that though, I'd say everything you need is already available, it's just a matter of maintenance burden (dealing with "magic string" route names isn't as friendly in the IDE as dealing with `Controller::class`).

**Reply by Pixelshaped** [reply]
Ha yes I forgot that it was a callable you are right.

**Comment by Pixelshaped** [comment]
Thanks for the help!
